diff_drive_controller/src/velocity_limiter.cpp

```diff
@@ -0,0 +1,146 @@
+    if (wheel_separation <= 0.0f) {
+        return false;
+    }
+    _wheel_separation = wheel_separation;
+    _has_velocity_limits = has_velocity_limits;
+    _has_acceleration_limits = has_acceleration_limits;
+
+    if (_has_acceleration_limits) {
+        if (acc_x_max <= 0 || acc_x_min >= 0) {
+            return false;
+        }
+
+        _f_wheel_max = 0.5f * _mass * acc_x_max;
+        _f_wheel_min = 0.5f * _mass * acc_x_min;
+
+        // J * acc_th_max = 2 * f_wheel_max / (wheel_separation / 2)
+        _inertia = _wheel_separation * _f_wheel_max / acc_th_max;
+    }
+
+    if (_has_velocity_limits) {
+        if (vel_x_max <= 0 || vel_th_max <= 0) {
+            return false;
+        }
+        _vel_x_max = vel_x_max;
+        _vel_th_max = vel_th_max;
+    }
+    return true;
+};
+
+void VelocityLimiter::limit(Vector& vel_cmd, const Vector& vel_cmd_prev, float dt) const {
+    // emergency stop
+    if (vel_cmd.x == 0.0f && vel_cmd.th == 0.0f) {
+        return;
+    }
+
+    if (_has_velocity_limits) {
+        limit_vel(vel_cmd);
+    }
+
+    if (_has_acceleration_limits) {
+        limit_acc(vel_cmd, vel_cmd_prev, dt);
+    }
+    return;
+};
+
+void VelocityLimiter::limit_vel(Vector& vel_cmd) const {
+    // initial down-scaling for angular velocity limit
+    downscale_both(vel_cmd.x, -_vel_x_max, _vel_x_max, vel_cmd.th, -_vel_th_max, _vel_th_max);
+
+    // convert to wheel velocities
+    float v_left = vel_cmd.x - vel_cmd.th * _wheel_separation * 0.5f;
+    float v_right = vel_cmd.x + vel_cmd.th * _wheel_separation * 0.5f;
+
+    // clip
+    downscale_both(v_left, v_right, -_vel_x_max, _vel_x_max);
+
+    // convert back to x/th velocities
+    vel_cmd.x = 0.5 * (v_left + v_right);
+    vel_cmd.th = (v_right - v_left) / _wheel_separation;
+}
+
+void VelocityLimiter::limit_acc(Vector& vel_cmd, const Vector& vel_cmd_prev, const float dt) const {
+    Vector acc{(vel_cmd.x - vel_cmd_prev.x) / dt, (vel_cmd.th - vel_cmd_prev.th) / dt};
+
+    // calculate wheel acceleration forces
+    float f_left = 0.5f * _mass * acc.x - _inertia * acc.th / _wheel_separation;
+    float f_right = 0.5f * _mass * acc.x + _inertia * acc.th / _wheel_separation;
+
+    // calculate previous wheel velocities
+    const float v_left_prev = vel_cmd_prev.x - vel_cmd_prev.th * _wheel_separation * 0.5f;
+    const float v_right_prev = vel_cmd_prev.x + vel_cmd_prev.th * _wheel_separation * 0.5f;
+
+    // decide if wheel speed is increasing (acceleration) or decreasing (deceleration)
+    const bool left_acc = f_left * v_left_prev >= 0.0f;
+    const bool right_acc = f_right * v_right_prev >= 0.0f;
+
+    // saturation magnitude
+    const float f_left_mag = left_acc ? _f_wheel_max : -_f_wheel_min;
+    const float f_right_mag = right_acc ? _f_wheel_max : -_f_wheel_min;
+
+    // limit
+    downscale_both(f_left, -f_left_mag, f_left_mag, f_right, -f_right_mag, f_right_mag);
```

Comment:
shouldn't the forward/backward acceleration limits happen in body frame instead of per wheel basis?